File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/savedcards/FetchSavedCards.java
package org.mifos.mobilewallet.core.domain.usecase.savedcards;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.entity.savedcards.Card;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.utils.Constants;

import java.util.List;

import javax.inject.Inject;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by ankur on 21/May/2018
 */

public class FetchSavedCards extends UseCase<FetchSavedCards.RequestValues,
        FetchSavedCards.ResponseValue> {

    private final FineractRepository mFineractRepository;

    @Inject
    public FetchSavedCards(FineractRepository fineractRepository) {
        mFineractRepository = fineractRepository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {
        mFineractRepository.fetchSavedCards(requestValues.clientId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<Card>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getUseCaseCallback().onError(e.toString());
                    }

                    @Override
                    public void onNext(List<Card> cards) {
                        if (cards != null) {
                            getUseCaseCallback().onSuccess(new
                                    ResponseValue(cards));
                        } else {
                            getUseCaseCallback().onError(Constants.NO_SAVED_CARDS);
                        }
                    }
                });
    }

    public static final class RequestValues implements UseCase.RequestValues {

        private final long clientId;

        public RequestValues(long clientId) {
            this.clientId = clientId;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        private final List<Card> cardList;

        public ResponseValue(List<Card> cards) {
            this.cardList = cards;
        }

        public List<Card> getCardList() {
            return cardList;
        }
    }
}

File: core/data/src/main/java/org/mifos/mobilewallet/core/domain/usecase/savedcards/FetchSavedCards.kt
package org.mifos.mobilewallet.core.domain.usecase.savedcards

import org.mifos.mobilewallet.core.base.UseCase
import com.mifos.mobilewallet.model.entity.savedcards.Card
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository
import org.mifos.mobilewallet.core.utils.Constants
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by ankur on 21/May/2018
 */
class FetchSavedCards @Inject constructor(private val mFineractRepository: FineractRepository) :
    UseCase<FetchSavedCards.RequestValues, FetchSavedCards.ResponseValue>() {

    class RequestValues(val clientId: Long) : UseCase.RequestValues
    class ResponseValue(val cardList: List<Card>) : UseCase.ResponseValue

    override fun executeUseCase(requestValues: RequestValues) {
        mFineractRepository.fetchSavedCards(requestValues.clientId)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<List<Card>>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    useCaseCallback.onError(e.toString())
                }

                override fun onNext(cards: List<Card>) {
                    if (cards.isNotEmpty()) {
                        useCaseCallback.onSuccess(ResponseValue(cards))
                    } else {
                        useCaseCallback.onError(Constants.NO_SAVED_CARDS)
                    }
                }
            })

    }
}
